### PR TITLE
Implement source module

### PR DESCRIPTION
**Pull request by niklaslong**
follow up to #191 

- [x] Set up mocks for Google and OpenAQ
- [x] Implement Sources module
- [x] Test sources module (mocking Google and OpenAQ)
- [x] add source flag
- [x] check for source before querying Google
- [ ] Link up to Beethe module
- [ ] Test Breethe module (perhaps functional tests rather than unit?)

Moved to Follow up:
- [ ] refactor by using `{lat, lon}` instead of `lat, lon`

### CODE REVIEW

**Comment by niklaslong** [comment]
@marcoow this implements a first version of the `Source` module. It isn't yet used by the `Breethe` module, but once everything is ready we'll link it all up. (see #104 for a plan and progress of the Eu data implementation)

The `lat, lon` refactor to follow will allow us to reduce a lot of code duplication, but it involves the whole code base which is why I didn't include it in this PR.

File: apps/breethe/lib/breethe/sources/google/geocoding.ex [hunk 1]
```diff
@@ -27,6 +32,46 @@ defmodule Breethe.Sources.Google.Geocoding do
...
+      })
+
+    query
+    |> query_google_api()
+    |> Jason.decode!()
+    |> (& &1["results"]).()
+    |> List.first()
+    |> (& &1["address_components"]).()
+    |> find_country_code()
+  end
```
Comment (marcoow) [inline apps/breethe/lib/breethe/sources/google/geocoding.ex]: I guess there is potential for refactoring these methods and extracting a shared helper method as both are almost identical?

Comment (niklaslong) [reply apps/breethe/lib/breethe/sources/google/geocoding.ex]: Yep, I wanted to include this in the `lat, lon` refactor.

**Comment by niklaslong** [comment]
@marcoow I've made some changes:
- If the cache **is not** empty, we use the `source` field to dictate the logic (either query OpenAQ or return the cache if it's from EEA) 
- If cache **is** empty, we have no choice but to query Google to find the location based on `search_term` so that we can figure out whether to query OpenAQ or return the cache if it's from EEA. In this case, speed isn't so much an issue as the first time we search for a location it's slow anyway.

**Comment by marcoow** [comment]
> If cache is empty, we have no choice but to query Google to find the location based on search_term so that we can figure out whether to query OpenAQ or return the cache if it's from EEA. In this case, speed isn't so much an issue as the first time we search for a location it's slow anyway.

Hm, I think in that case we don't need to query Google at all though as we know we will definitely have to get the data from OpenAQ anyway? So I think you can make this shorter:

```
cache empty -> query OpenAQ
```

File: apps/breethe/lib/breethe/sources/google/geocoding.ex (hunk 1 above)
Comment (niklaslong) [reply apps/breethe/lib/breethe/sources/google/geocoding.ex]: Sure, I can do it here but it involves the whole code base which is why I thought it best to split it up?

**Comment by niklaslong** [comment]
> > If cache is empty, we have no choice but to query Google to find the location based on search_term so that we can figure out whether to query OpenAQ or return the cache if it's from EEA. In this case, speed isn't so much an issue as the first time we search for a location it's slow anyway.
> 
> Hm, I think in that case we don't need to query Google at all though as we know we will definitely have to get the data from OpenAQ anyway? So I think you can make this shorter:
> 
> ```
> cache empty -> query OpenAQ
> ```

I assumed we didn't want to use OpenAQ as a fallback even if the European data is empty?

**Comment by niklaslong** [comment]
> I think I might be missing something here – is the new module not actually used anywhere?

This is true - again to avoid a massive PR it's not being used by the `Breethe` module yet (a rewrite of the tests will be necessary). But again if you think it's best to include everything in this one rather than doing the implementation in steps, I can do that?

**Comment by marcoow** [comment]
> I assumed we didn't want to use OpenAQ as a fallback even if the European data is empty?

Yes, that's the goal but if we're introducing the EEA data side-by-side the OpenAQ data here, we are actually using both at the same time?

> This is true - again to avoid a massive PR it's not being used by the Breethe module yet (a rewrite of the tests will be necessary). But again if you think it's best to include everything in this one rather than doing the implementation in steps, I can do that?

Yes, we should make this PR so that the code that is introduced is actually used. Otherwise, we cannot know whether it actually works the way we want it anyway.

File: apps/breethe/lib/breethe/sources/google/geocoding.ex (hunk 1 above)
Comment (niklaslong) [reply apps/breethe/lib/breethe/sources/google/geocoding.ex]: Ah right - yes I'd misunderstood, sorry about that!
fixed

**Comment by niklaslong** [comment]
replaced by #203

**Pull request closed without merging**